### ISSUE TITLE
eeg tech. inbound  with 50 hz

### ISSUE THREAD

**Issue by 9535299035** [9535299035/eeg-monitor#6]
import numpy as np

from scipy.signal import butter, filtfilt, iirnotch, welch
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler

class EEG_BCI:

    def __init__(self, sampling_rate=250):

        self.fs = sampling_rate

        self.scaler = StandardScaler()

        self.classifier = LinearDiscriminantAnalysis()

        self.trained = False

    # ------------------------------------------------
    # FILTER
    # ------------------------------------------------

    def bandpass(self, eeg):

        nyquist = self.fs / 2

        b, a = butter(
            4,
            [1 / nyquist, 40 / nyquist],
            btype="band"
        )

        return filtfilt(
            b,
            a,
            eeg
        )

    # ------------------------------------------------
    # REMOVE 50 Hz NOISE
    # ------------------------------------------------

    def notch(self, eeg):

        b, a = iirnotch(
            50,
            30,
            self.fs
        )

        return filtfilt(
            b,
            a,
            eeg
        )

    # ------------------------------------------------
    # PREPROCESS EEG
    # ------------------------------------------------

    def preprocess(self, eeg):

        eeg = np.asarray(
            eeg,
            dtype=float
        )

        # Remove DC component
        eeg = eeg - np.mean(
            eeg,
            axis=-1,
            keepdims=True
        )

        # Bandpass
        eeg = self.bandpass(
            eeg
        )

        # Notch
        eeg = self.notch(
            eeg
        )

        return eeg

    # ------------------------------------------------
    # FREQUENCY FEATURES
    # ------------------------------------------------

    def frequency_features(
        self,
        eeg
    ):

        frequencies, power = welch(
            eeg,
            fs=self.fs,
            axis=-1,
            nperseg=min(
                eeg.shape[-1],
                256
            )
        )

        bands = {

            "delta": (1, 4),

            "theta": (4, 8),

            "alpha": (8, 13),

            "beta": (13, 30),

            "gamma": (30, 40)
        }

        features = []

        for low, high in bands.values():

            mask = (
                (frequencies >= low) &
                (frequencies < high)
            )

            band_power = np.trapezoid(
                power[..., mask],
                frequencies[mask],
                axis=-1
            )

            features.append(
                band_power
            )

        return np.concatenate(
            features,
            axis=-1
        )

    # ------------------------------------------------
    # FEATURE EXTRACTION
    # ------------------------------------------------

    def extract_features(self, eeg):

        cleaned = self.preprocess(
            eeg
        )

        features = self.frequency_features(
            cleaned
        )

        return features

    # ------------------------------------------------
    # TRAIN BCI
    # ------------------------------------------------

    def train(
        self,
        eeg_training_data,
        labels
    ):

        feature_data = []

        for eeg in eeg_training_data:

            features = self.extract_features(
                eeg
            )

            feature_data.append(
                features
            )

        X = np.asarray(
            feature_data
        )

        X = self.scaler.fit_transform(
            X
        )

        self.classifier.fit(
            X,
            labels
        )

        self.trained = True

        print(
            "BCI training complete."
        )

    # ------------------------------------------------
    # PREDICT
    # ------------------------------------------------

    def predict(self, eeg):

        if not self.trained:

            raise RuntimeError(
                "BCI has not been trained."
            )

        features = self.extract_features(
            eeg
        )

        features = self.scaler.transform(
            features.reshape(1, -1)
        )

        prediction = self.classifier.predict(
            features
        )[0]

        probability = self.classifier.predict_proba(
            features
        )[0]

        confidence = np.max(
            probability
        )

        return prediction, confidence

    # ------------------------------------------------
    # COMMAND DECODER
    # ------------------------------------------------

    def command(self, prediction, confidence):

        if confidence < 0.70:

            return "NO COMMAND"

        commands = {

            0: "LEFT",

            1: "RIGHT",

            2: "FORWARD",

            3: "BACKWARD",

            4: "STOP"
        }

        return commands.get(
            prediction,
            "UNKNOWN"
        )
